sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -33,13 +33,24 @@ case class CreateArray(children: Seq[Expression]) extends Expression {
 
   override def foldable: Boolean = children.forall(_.foldable)
 
-  override def checkInputDataTypes(): TypeCheckResult =
-    TypeUtils.checkForSameTypeInputExpr(children.map(_.dataType), "function array")
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (children.map(_.dataType).forall(_.isInstanceOf[DecimalType])) {
+      TypeCheckResult.TypeCheckSuccess
+    } else {
+      TypeUtils.checkForSameTypeInputExpr(children.map(_.dataType), "function array")
+    }
+  }
 
   override def dataType: DataType = {
-    ArrayType(
-      children.headOption.map(_.dataType).getOrElse(NullType),
-      containsNull = children.exists(_.nullable))
+    var elementType: DataType = children.headOption.map(_.dataType).getOrElse(NullType)
+    if (elementType.isInstanceOf[DecimalType]) {
+      children.foreach { child =>
+        if (elementType.asInstanceOf[DecimalType].isTighterThan(child.dataType)) {
```

Comment:
Thank you, @rxin .
Yep. I've read you comment about the lose.
I'll check that and revise.